### PR TITLE
Optimize Presto Serialization

### PR DESCRIPTION
**Pull request by oerling**
Uses SIMD and uses ranges of indices instead of ranges of index intervals in vector serializers.

### CODE REVIEW

File: velox/serializers/PrestoSerializer.cpp [hunk 1]
```diff
@@ -2024,11 +2709,308 @@ void estimateSerializedSizeInt(
...
+    for (auto i = 0; i < rows.size(); ++i) {
+      *sizes[i] += rawValues[rows[i]].size();
+    }
+  } else {
+    ScratchPtr<uint64_t, 4> nullsHolder(scratch);
+    ScratchPtr<int32_t, 64> nonNullsHolder(scratch);
+    auto nulls = nullsHolder.get(bits::nwords(numRows));
+    simd::gatherBits(rawNulls, rows, nulls);
+    auto* nonNulls = nonNullsHolder.get(numRows);
+    auto numNonNull = simd::indicesOfSetBits(nulls, 0, numRows, nonNulls);
```
Comment (Yuhta) [inline velox/serializers/PrestoSerializer.cpp:2775]: `forEachSetBit`

File: velox/serializers/PrestoSerializer.cpp [hunk 2]
```diff
@@ -2024,11 +2709,308 @@ void estimateSerializedSizeInt(
...
+          innerSizes[i] = sizes[mutableInnerRows[i]];
+        }
+        simd::transpose(
+            rows.data(),
+            folly::Range<const vector_size_t*>(mutableInnerRows, numInner),
+            mutableInnerRows);
+        innerRows = mutableInnerRows;
+      }
+      auto rowVector = vector->as<RowVector>();
+      auto children = rowVector->children();
```
Comment (Yuhta) [inline velox/serializers/PrestoSerializer.cpp:2942]: `auto& children`

File: velox/serializers/PrestoSerializer.cpp [hunk 3]
```diff
@@ -2024,11 +2709,308 @@ void estimateSerializedSizeInt(
...
+          scratch);
+      if (numRanges == 0) {
+        return;
+      }
+      estimateSerializedSizeInt(
+          arrayVector->elements().get(),
+          folly::Range<const IndexRange*>(rangeHolder.get(), numRanges),
+          sizesHolder.get(),
+          scratch);
+      estimateSerializedSizeInt(
```
Comment (Yuhta) [inline velox/serializers/PrestoSerializer.cpp:3005]: Why do it twice?

File: velox/exec/PartitionedOutput.cpp
Comment (Yuhta) [inline velox/exec/PartitionedOutput.cpp:1]: Let's leave out the operator change to next PR

**Comment by oerling** [comment]
From: Jimmy Lu ***@***.***> 
Sent: Wednesday, November 29, 2023 7:05 PM
To: facebookincubator/velox  

+
+  if constexpr (sizeof(T) == 8) {

No need to branch on sizeof(T)

A: Less duplication. If special simd for 32 and 64, this is useful.

 

 

 

  _____  

In velox/serializers/PrestoSerializer.cpp <https://github.com/facebookincubator/velox/pull/7565#discussion_r1409831157> :

> +    uint8_t nullsByte = *reinterpret_cast<const uint8_t*>(nulls);
+    numNonNull = __builtin_popcount(nullsByte);
+    nonNullIndices =
+        numNonNull == numRows ? nullptr : simd::byteSetBits(nullsByte);
+  } else {
+    auto mutableIndices = nonNullHolder.get(nu utput = window.get(numNonNull);
+    if (numNonNull == numRows) {

Remove unused code

 

 

A: Not at all. There can be nulls in the column and there can be no nulls in th selected range, so the switch is legit.  Less indirection if all indices are in. 

 

 

  _____  

In velox/serializers/PrestoSerializer.cpp <https://github.com/facebookincubator/velox/pull/7565#discussion_r1409831332> :

> +    folly::Range<const vector_size_t*> rows,
+    const StringView* views,
+    VectorStream* stream,
+    Scratch& scratch) {
+  if (!nulls) {
+    stream->appendLengths(nullptr, rows, rows.size(), [&](auto row) {
+      return views[row].size();
+    });
+    for (auto i = 0; i < rows.size(); ++i) {
+      auto& view = views[rows[i]];
+      stream->values().appendStringView(
+          std::string_view(view.data(), view.size()));
+    }
+    return;
+  }
+  auto nonNull = rows.data();

Do not initialize these two here with rows containing nulls

 

A: OK

 

 

 

  _____  

In velox/serializers/PrestoSerializer.cpp <https://github.com/facebookincubator/velox/pull/7565#discussion_r1409831606> :

> +  auto nonNullRows = nonNullHolder.get(rows.size());
+  auto numNonNull = simd::indicesOfSetBits(nulls, 0, rows.size(), nonNullRows);
+  stream->appendNulls(nulls, 0, rows.size(), numNonNull);
+  for (auto i = 0; i < numNonNull; ++i) {
+    stream->appendOne(timestamps[rows[nonNullRows[i]]]);
+  }
+}
+
+template <TypeKind kind>
+void serializeFlatVector(
+    const BaseVector* vector,
+    const folly::Range<const vector_size_t*>& rows,
+    VectorStream* stream,
+    Scratch& scratch) {
+  using T = typename TypeTraits<kind>::NativeType;
+  auto flatVector = reinterpret_cast<const FlatVector<T>*>(vector);

vector->asFlatVector<T>()

 

A: asUnchecked. As is dynamic_cast. Bad. 

 

 

  _____  

In velox/serializers/PrestoSerializer.cpp <https://github.com/facebookincubator/velox/pull/7565#discussion_r1410080372> :

> @@ -1710,6 +1801,606 @@ void serializeColumn(
   }
 }
 
+// Returns ranges for the non-null rows of an array  or map. 'rows' gives the
+// rows. nulls is the nulls of the array/map or nullptr if no nulls. 'offsets'
+// and 'sizes' are the offsets and sizes of the array/map.Returns the number of
+// index ranges. Obtains the ranges from 'rangesHolder'. If 'sizesPtr' is
+// non-null, gets returns  the sizes for the inner ranges in 'sizesHolder'. If
+// 'stream' is non-null, writes the lengths and nulls for the array/map into
+// 'stream'.
+int32_t rowsToRanges(
+    folly::Range<const vector_size_t*> rows,
+    const uint64_t* rawNulls,
+    const vector_size_t* offsets,
+    const vector_size_t* sizes,
+    vector_size_t** sizesPtr,
+    ScratchPtr<IndexRange>& rangesHolder,

Should pass raw pointers instead of passing ScratchPtr

 

 

A: No.  ScratchPtr will be conditionally referenced. The size is not known outside of the call. 

 

 

  _____  

In velox/serializers/PrestoSerializer.cpp <https://github.com/facebookincubator/velox/pull/7565#discussion_r1410080866> :

> +  }
+
+  T value = constVector->valueAtFast(0);
+  for (int32_t i = 0; i < numRows; ++i) {
+    stream->appendNonNull();
+    stream->appendOne(value);
+  }
+}
+
+template <typename T>
+void serializeBiasVector(
+    const BaseVector* vector,
+    const folly::Range<const vector_size_t*>& rows,
+    VectorStream* stream,
+    Scratch& scratch) {
+  VELOX_UNSUPPORTED()

Missing semicolon

  _____  

On velox/serializers/tests/PrestoSerializerTest.cpp <https://github.com/facebookincubator/velox/pull/7565#discussion_r1410089057> :

In testRoundTrip, let's test the vector as top level column, and wrap it inside an array, then test it again

  _____  

On velox/serializers/PrestoSerializer.cpp <https://github.com/facebookincubator/velox/pull/7565#discussion_r1410089606> :

Can we find some way to reduce the code duplication? Now every function has a version taking ranges and rows, is there a way to abstract this out with a template?

 

A: Not in a  practical way. The range variants can be optimized further but that is later. 

 

 

  _____  

In velox/serializers/PrestoSerializer.cpp <https://github.com/facebookincubator/velox/pull/7565#discussion_r1410092877> :

> +template <TypeKind Kind>
+void estimateFlatSerializedSize(
+    const BaseVector* vector,
+    const folly::Range<const vector_size_t*>& rows,
+    vector_size_t** sizes,
+    Scratch& scratch) {
+  const auto valueSize = vector->type()->cppSizeInBytes();
+  const auto numRows = rows.size();
+  if (vector->mayHaveNulls()) {
+    auto rawNulls = vector->rawNulls();
+    ScratchPtr<uint64_t, 4> nullsHolder(scratch);
+    ScratchPtr<int32_t, 64> nonNullsHolder(scratch);
+    auto nulls = nullsHolder.get(bits::nwords(numRows));
+    simd::gatherBits(rawNulls, rows, nulls);
+    auto nonNulls = nonNullsHolder.get(numRows);
+    const auto numNonNull = simd::indicesOfSetBits(nulls, 0, numRows, nonNulls);

forEachSetBit?

 

 

A: No, we expect mostly non-null. Need loop over contiguous and simd write of mostly 8 positions.

 

 

  _____  

In velox/serializers/PrestoSerializer.cpp <https://github.com/facebookincubator/velox/pull/7565#discussion_r1410100820> :

> +    Scratch& scratch) {
+  const auto numRows = rows.size();
+  auto strings = static_cast<const FlatVector<StringView>*>(vector);
+  auto rawNulls = strings->rawNulls();
+  auto rawValues = strings->rawValues();
+  if (!rawNulls) {
+    for (auto i = 0; i < rows.size(); ++i) {
+      *sizes[i] += rawValues[rows[i]].size();
+    }
+  } else {
+    ScratchPtr<uint64_t, 4> nullsHolder(scratch);
+    ScratchPtr<int32_t, 64> nonNullsHolder(scratch);
+    auto nulls = nullsHolder.get(bits::nwords(numRows));
+    simd::gatherBits(rawNulls, rows, nulls);
+    auto* nonNulls = nonNullsHolder.get(numRows);
+    auto numNonNull = simd::indicesOfSetBits(nulls, 0, numRows, nonNulls);

forEachSetBit

 

A: No

 

  _____  

In velox/serializers/PrestoSerializer.cpp <https://github.com/facebookincubator/velox/pull/7565#discussion_r1410101368> :

> +        auto nulls = nullsHolder.get(bits::nwords(numRows));
+        simd::gatherBits(vector->rawNulls(), rows, nulls);
+        auto mutableInnerRows = innerRowsHolder.get(numRows);
+        numInner = simd::indicesOfSetBits(nulls, 0, numRows, mutableInnerRows);
+        innerSizes = innerSizesHolder.get(numInner);
+        for (auto i = 0; i < numInner; ++i) {
+          innerSizes[i] = sizes[mutableInnerRows[i]];
+        }
+        simd::transpose(
+            rows.data(),
+            folly::Range<const vector_size_t*>(mutableInnerRows, numInner),
+            mutableInnerRows);
+        innerRows = mutableInnerRows;
+      }
+      auto rowVector = vector->as<RowVector>();
+      auto children = rowVector->children();

auto& children

 

A: Good point

 

  _____  

In velox/serializers/PrestoSerializer.cpp <https://github.com/facebookincubator/velox/pull/7565#discussion_r1410101957> :

> +          arrayVector->rawOffsets(),
+          arrayVector->rawSizes(),
+          sizes,
+          rangeHolder,
+          &sizesHolder,
+          nullptr,
+          scratch);
+      if (numRanges == 0) {
+        return;
+      }
+      estimateSerializedSizeInt(
+          arrayVector->elements().get(),
+          folly::Range<const IndexRange*>(rangeHolder.get(), numRanges),
+          sizesHolder.get(),
+          scratch);
+      estimateSerializedSizeInt(

Why do it twice?

 

A: Deleted

 

  _____  

On velox/exec/PartitionedOutput.cpp <https://github.com/facebookincubator/velox/pull/7565#discussion_r1410102737> :

Let's leave out the operator change to next PR

**Comment by Yuhta** [comment]
New PR #7883

**Pull request closed without merging**